### PR TITLE
CAY-2853 Incorrect deletion of entities from flattened attributes

### PR DESCRIPTION
**Pull request by Jugen**
When deleting an ObjEntity with flattened attributes Cayenne deletes the related DB attribute's row as well.
 
So if we have for example an ObjEntity "FlatPainting" with flattened attribute "galleryName". If we then delete a "FlatPainting" then the associated gallery will also be deleted, leaving all other paintings at that gallery without a valid Gallery link.
 
Furthermore if we add to our "FlatPainting" a "gallery" relationship attribute with delete rule Deny, the gallery was still being deleted.

The original intention was for something like a "PaintingFull" ObjEntity having all painting attributes and a flattened attribute to "PaintingInfo" as well, where the relationship is 1:1. In this case it makes sense that deleting a "PaintingFull" should also delete its associated PaintingInfo record.

In this PR a check has been added so that reverse toMany relationships are protected from deletion,
and if a relevant ObjRelationship is present then deletions are deferred to it instead.

### CODE REVIEW

**Comment by Jugen** [comment]
This PR is ready for review please.

File: cayenne/src/main/java/org/apache/cayenne/access/ObjectStore.java [hunk 1]
```diff
@@ -1011,6 +1013,18 @@ public Collection<ObjectId> getFlattenedIds(ObjectId objectId) {
                 .getOrDefault(objectId, Collections.emptyMap()).values();
     }
 
+    /**
+     * @since 5.0
+     */
+    public Set<Entry<CayennePath,ObjectId>> getFlattenedEntries(ObjectId objectId) {
```
Comment (stariy95) [inline cayenne/src/main/java/org/apache/cayenne/access/ObjectStore.java:1019]: I think `Map<CayennePath, ObjectId>` would be a bit cleaner to return here.

File: cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java [hunk 2]
```diff
@@ -74,14 +79,37 @@ public Void visitUpdate(UpdateDbRowOp dbRow) {
...
+
+        flattenedEntries.forEach(entry -> {
+            DbRelationship dbRel = dbSource.getRelationship(entry.getKey().first().toString());
+
+            // Don't delete if the target entity has a toMany relationship with the source entity,
+            // as there may be other records in the source entity with references to it.
+            if (!dbRel.getReverseRelationship().isToMany()) {
+
+                // Check if there's an ObjRelationship matching the flattened attributes DbRelationship
+                boolean hasObjRelationship = entity.getRelationships().stream()
```
Comment (stariy95) [inline cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java:100]: This looks heavy, though not many flattened attributes are usually mapped. Ultimately this could be precalculated and moved to the model (`ClassDescriptor.additionalDbEntities` maybe?), but that is definitely 5.0 task only, while this fix could be easily backported to 4.2

File: cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java [hunk 3]
```diff
@@ -457,28 +457,42 @@ public void testDelete() throws Exception {
...
+                .select(runtime.newContext());
+        
+        // Should now only have one painting by artist2
+        assertEquals("Painting", 1, objects.size());
+        // and that painting should have a valid gallery
+        assertNotNull("Gallery is null", objects.get(0).getToGallery());
+        assertNotNull("GalleryName is null", objects.get(0).getToGallery().getGalleryName());
+        
+        // There should be one less painting info now
+        Number infoCount = (Number) Cayenne.objectForQuery(context, new EJBQLQuery(
```
Comment (stariy95) [inline cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java:493]: A minor feature advertisement, you could query it just like this: `long count = ObjectSelect.query(PaintingInfo.class).selectCount(context);`

File: cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java [hunk 4]
```diff
@@ -457,28 +457,42 @@ public void testDelete() throws Exception {
         a.setArtistName("AX");
         context.commitChanges();
 
-        CompoundPainting o1 = context.newObject(CompoundPainting.class);
-        o1.setArtistName("A1");
-        o1.setEstimatedPrice(new BigDecimal(1.0d));
-        o1.setGalleryName("G1");
-        o1.setPaintingTitle("P1");
-        o1.setTextReview("T1");
+        createTestDataSet();
```
Comment (stariy95) [inline cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java:460]: I think it's better to keep an old test as is and just add a new one for the task

**Review by stariy95 — COMMENTED** [review]
@Jugen Thanks for the PR! I got some thoughts on how we may improve it a bit.

File: cayenne/src/main/java/org/apache/cayenne/access/ObjectStore.java (hunk 1 above)
Comment (Jugen) [reply cayenne/src/main/java/org/apache/cayenne/access/ObjectStore.java:1019]: I thought so too, but didn't know how to name the method then as `getFlattenedMap` didn't seem appropriate.
Any suggestions please ?

File: cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java (hunk 3 above)
Comment (Jugen) [reply cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java:493]: Good idea, will do.

File: cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java (hunk 4 above)
Comment (Jugen) [reply cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java:460]: I replaced it because the old test was faulty. If you prefer I can restore it with corrections though ?

File: cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java (hunk 2 above)
Comment (Jugen) [reply cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java:100]: I'd prefer this being backported to 4.2 if possible.

Comment (stariy95) [reply cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java:100]: Sure, let's keep it as is for 4.2 compatibility.

File: cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java (hunk 4 above)
Comment (stariy95) [reply cayenne/src/test/java/org/apache/cayenne/access/DataContextFlattenedAttributesIT.java:460]: Yeah, I think we should fix it. There's always shortage of the tests for corner cases.

**Comment by Jugen** [comment]
Have made all the changes as requested. Note that the changes to `RootRowOpProcessor.visitDelete` has been slightly modified from the initial PR in order for the restored test to pass. So that was a good thing :-). It's strange though .....

**Review by stariy95 — APPROVED** [review]

**Comment by Jugen** [comment]
Updated release notes and squashed commits for this PR.

**Comment by Jugen** [comment]
Now that this has been merged are there steps that I must take to back port this as previously suggested ?

**Comment by stariy95** [comment]
@Jugen I'll try to just cherry-pick this commit, and hope that it'll just apply.
